File: src/main/java/com/simon/java/example/ProducerConsumerExample.java
package com.simon.java.example;

/**
 * @author Simon
 * @desc
 */

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

public class ProducerConsumerExample {
    public static void main(String[] args) {
        BlockingQueue<Integer> queue = new ArrayBlockingQueue<>(10);
        Thread producer = new Thread(new Producer(queue));
        Thread consumer = new Thread(new Consumer(queue));
        producer.start();
        consumer.start();
    }

    static class Producer implements Runnable {
        private final BlockingQueue<Integer> queue;

        public Producer(BlockingQueue<Integer> queue) {
            this.queue = queue;
        }

        @Override
        public void run() {
            try {
                for (int i = 0; i < 10; i++) {
                    System.out.println("Producing " + i);
                    Thread.sleep(1000);
                    queue.put(i);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    static class Consumer implements Runnable {
        private final BlockingQueue<Integer> queue;

        public Consumer(BlockingQueue<Integer> queue) {
            this.queue = queue;
        }

        @Override
        public void run() {
            try {
                for (int i = 0; i < 10; i++) {
                    Integer value = queue.take();
                    System.out.println("Consuming " + value);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/main/kotlin/com/simon/kotlin/example/ProducerConsumerExample.kt
package com.simon.kotlin.example

/**
 * @author Simon
 * @desc
 */
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.*

fun main() {
    runBlocking {
        val channel = Channel<Int>(10)
        val producer = launch {
            for (i in 0..9) {
                println("Producing $i")
                delay(1000)
                channel.send(i)
            }
        }
        val consumer = launch {
            repeat(10) {
                val value = channel.receive()
                println("Consuming $value")
            }
        }
        producer.join()
        consumer.join()
    }
}
